shared/AppInsightsCore/src/JavaScriptSDK/TelemetryInitializerPlugin.ts

```diff
@@ -0,0 +1,119 @@
+}
+
+export class TelemetryInitializerPlugin extends BaseTelemetryPlugin implements ITelemetryInitializerContainer {
+
+    public identifier: string = "TelemetryInitializerPlugin";
+    priority: number = 199;
+
+    constructor() {
+        super();
+
+        // NOTE!: DON'T set default values here, instead set them in the _initDefaults() function as it is also called during teardown()
+        let _id: number;
+        let _initializers: _IInternalTelemetryInitializerHandler[];
+
+        _initDefaults();
+
+        dynamicProto(TelemetryInitializerPlugin, this, (_self, _base) => {
+
+            _self.addTelemetryInitializer = (telemetryInitializer: TelemetryInitializerFunction): ITelemetryInitializerHandler => {
+                let theInitializer = {
+                    id: _id++,
+                    fn: telemetryInitializer
+                };
+
+                _initializers.push(theInitializer);
+
+                let handler: ITelemetryInitializerHandler = {
+                    remove: () => {
+                        arrForEach(_initializers, (initializer, idx) => {
+                            if (initializer.id === theInitializer.id) {
+                                _initializers.slice(idx, 1);
+                                return -1;
+                            }
+                        });
+                    }
+                }
+    
+                return handler;
+            }
+
+            _self.processTelemetry = (item: ITelemetryItem, itemCtx?: IProcessTelemetryContext): void => {
+                var doNotSendItem = false;
+                var telemetryInitializersCount = _initializers.length;
+                for (var i = 0; i < telemetryInitializersCount; ++i) {
+                    var telemetryInitializer = _initializers[i];
+                    if (telemetryInitializer) {
+                        try {
+                            if (telemetryInitializer.fn.apply(null, [item]) === false) {
+                                doNotSendItem = true;
+                                break;
+                            }
+                        } catch (e) {
+                            // log error but dont stop executing rest of the telemetry initializers
+                            // doNotSendItem = true;
+                            itemCtx.diagLog().throwInternal(
+                                LoggingSeverity.CRITICAL,
+                                _InternalMessageId.TelemetryInitializerFailed,
+                                "One of telemetry initializers failed, telemetry item will not be sent: " + getExceptionName(e),
+                                { exception: dumpObj(e) }, true);
+                        }
+                    }
+                }
+
+                if (!doNotSendItem) {
+                    _self.processNext(item, itemCtx);
+                }
+            };
+
+            _self.unload = (itemCtx: IProcessTelemetryContext, isAsync: boolean): void => {
```

Comment:
Ok, this is the signature of the new `unload` function that each plugin will be able to implement.
I thought I'd removed all of them from this PR (in an attempt to keep this PR smaller).